diffFiles/src/diff_files.py

```diff
@@ -1,106 +1,130 @@
+from dataclasses import dataclass
+from typing import cast, IO
 import click
 
 
+@dataclass
+class Position:
+    x: int
+    y: int
+
+
+@dataclass
 class Node:
-    def __init__(self, next_pos: tuple[int, int] | None, length: int) -> None:
-        self.next_pos = next_pos
-        self.length = length
+    next_pos: Position
+    length: int
 
 
-def build_subsequence_matrix(l1: list[str], l2: list[str]) -> list[list[Node]]:
-    # we use the matrix as a graph
-    # where each node points to one of their neighbors
-    nodes: list[list[Node]] = []
-
-    def get_node(pos: tuple[int, int]) -> Node | None:
-        return nodes[pos[0]][pos[1]] if pos[0] >= 0 and pos[1] >= 0 else None
-
-    def compute_node(i1: int, i2: int) -> Node:
-        next_pos: tuple[int, int] | None
-        length: int
-
-        if l1[i1] == l2[i2]:
-            next_pos = (i1 - 1, i2 - 1)
-            length = nodes[next_pos[0]][next_pos[1]].length + 1
-        else:
-            candidate1 = (i1, i2 - 1)
-            candidate2 = (i1 - 1, i2)
-            node1 = get_node(candidate1)
-            node2 = get_node(candidate2)
-
-            # find the candidate with the bigger length
-            if node1 is None and node2 is None:
-                next_pos = None
-                length = 0
-            elif node2 is not None and (node1 is None or node1.length < node2.length):
-                next_pos = candidate2
-                length = node2.length
-            else:
-                next_pos = candidate1
-                length = cast(Node, node1).length
+@dataclass
+class Matrix:
+    nodes: list[list[Node]]
 
-        return Node(next_pos, length)
+    def get_node(self, pos: Position) -> Node | None:
+        if pos.x < 0 or pos.y < 0:
+            return None
+        return self.nodes[pos.x][pos.y]
 
-    for i1 in range(len(l1)):
-        nodes.append([])
-        for i2 in range(len(l2)):
-            nodes[i1].append(compute_node(i1, i2))
+    @classmethod
+    def make(cls, l1: list[str], l2: list[str]) -> "Matrix":
+        def compute_node(i1: int, i2: int, line1, line2) -> Node:
+            next_pos: Position | None
+            length: int
+
+            if line1 == line2:
+                next_pos = Position(i1 - 1, i2 - 1)
+                length = result.nodes[next_pos.x][next_pos.y].length + 1
+            else:
+                candidate1 = Position(i1, i2 - 1)
+                candidate2 = Position(i1 - 1, i2)
+                node1 = result.get_node(candidate1)
+                node2 = result.get_node(candidate2)
+
+                # find the candidate with the bigger length
+                if node1 is None and node2 is None:
+                    next_pos = None
+                    length = 0
+                elif node2 is not None and (node1 is None or node1.length < node2.length):
+                    next_pos = candidate2
+                    length = node2.length
+                else:
+                    next_pos = candidate1
```

Comment:
In order for `next_pos.y = -1` we would need to enter this branch with `node1 is None` (as there is a guard inside get_node). However, if `node1 is None`, then either branch 1 or 2 will be entered depending on `node1 is None`.

So im not seeing what should be wrong here.